File: src/p2164.cpp
#include <stdio.h>
#include <queue>

int main(){
  int num;
  scanf("%d", &num);

  std::queue<int> q;
  // add to queue
  for (int n = 1; n<=num; n++){
    q.push(n);
  }

  while(q.size() != 1) {
    q.pop(); // drop
    if(q.size() == 1)
      break;

    int top = q.front();
    q.pop();
    q.push(top);
  }

  printf("%d\n", q.front());

  return 0;
}

File: src/p2164.ml
exception Early_exit

let solve n =
  let q = Queue.create () in
  for num = 1 to n do
    Queue.push num q
  done;

  (try
     while Queue.length q != 1 do
       ignore (Queue.pop q);
       if Queue.length q = 1 then raise Early_exit;
       let top = Queue.take q in
       Queue.push top q
     done
   with
  | Early_exit -> ()
  | _ -> ());
  Printf.printf "%d\n" (Queue.top q)


let () = Scanf.scanf "%d" (fun n -> solve n)
